### PR TITLE
CA-325582: remove unused Host datasources to avoid space leaks

### PR DESCRIPTION
**Pull request by edwintorok**
This needs to be deployed with corresponding commit in rrdd-plugins.

Previously we always kept all datasources we've ever seen for a host
even if we temporarily stopped receiving RRD updates for it.
This is useful for per-host SR-level RRDs which only exist when at least
one VM using it is running: we want to see the historic data when the VM
is started again.
However if SRs are attached/detached frequently this can cause xcp-rrdd
to increase in memory usage over time, even across toolstack restarts or
host reboots, because all host-level RRDs were preserved.
This was not visible in `xe host-data-source-list`, since that lists
only the active RRDs.

Solution is to preserve only the latest reported RRDs for hosts, and to
ask plugins to keep reporting data for inactive RRDs that they need.
E.g. rrdp-iostat should keep reporting data for SRs that have a PBD for
that host, even if no VM is using it or the PBD is temporarily
unplugged.

Corressponding change in rrdp-iostat is WiP.

### CODE REVIEW

File: rrdd/rrdd_monitor.ml [hunk 1]
```diff
@@ -40,6 +40,17 @@ let merge_new_dss rrd dss =
       rrd_add_ds rrd now (Rrd.ds_create ds.ds_name ds.Ds.ds_type ~mrhb:300.0 Rrd.VT_Unknown)
     ) rrd new_dss
 
+module StringSet = Set.Make(String)
+
+let remove_unused_dss rrdi =
+  let all_dss = rrdi.rrd |> Rrd.ds_names in
+  let used_dss = rrdi.dss |> List.map (fun d -> d.ds_name) in
+  if List.length all_dss = List.length used_dss then rrdi
```
Comment (lindig) [inline rrdd/rrdd_monitor.ml]: Is this optimisation worth it? We need to travers at least both list. There is also the theoretical problem that both lists could have equal length but contain different names. The optimisation does save allocation of short-lived memory, though.

Comment (edwintorok) [reply rrdd/rrdd_monitor.ml]: Most of the time there is nothing to remove, so we can avoid building the stringsets.
IIUC the RRD would contain at least the entries from the DSS, possibly more, but perhaps we shouldn't rely on that detail here.

**Comment by lindig** [comment]
This looks ready to be merged but depends on other changes. Any comments from @psafont?

**Comment by psafont** [comment]
I haven't had time to test the changes and see the long-term impact this might do to metrics.

**Comment by psafont** [comment]
I think this PR is superseded by https://github.com/xapi-project/xen-api/pull/3989 and https://github.com/xapi-project/xcp-rrdd/pull/106

While this one has has the advantage that it cleans up all inactive datasources, it might be too aggressive and needs plugins to do extra effort.

With the other PRs it is now possible to observe the inactive sources from the xapi cli and remove them if there are more leaks, more fixes can be done to patch the leaks automatically in xapi if they are observed.

**Comment by lindig** [comment]
In light of the discussion above, should this PR be closed @edwintorok?

**Pull request closed without merging**